### PR TITLE
Support MSC4040 `matrix-fed` service name lookup

### PR DESCRIPTION
**Pull request by turt2live**
See https://github.com/matrix-org/matrix-spec/pull/1624

**Requires https://github.com/matrix-org/gomatrixserverlib/pull/411**

Domain tests:

<details>
<summary><code>3c.s.resolvematrix.dev</code> - <code>_matrix</code> only, <code>.well-known</code></summary>

image

</details>

<details>
<summary><code>4.s.resolvematrix.dev</code> - <code>_matrix</code> only</summary>

image

</details>

<details>
<summary><code>3c.msc4040.s.resolvematrix.dev</code> - <code>_matrix-fed</code> only, <code>.well-known</code></summary>

image

</details>

<details>
<summary><code>4.msc4040.s.resolvematrix.dev</code> - <code>_matrix-fed</code> only</summary>

image

</details>

<details>
<summary><code>t2l.io</code> - <code>_matrix-fed</code> and <code>_matrix</code> fallback</summary>

(TODO: Update resolvematrix.dev to support this case)

image

</details>

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -598,6 +628,25 @@ func (report *ServerReport) touchUpReport() {
...
+	// Check certificate chain validity
+	verificationOpts := x509.VerifyOptions{
+		// Certificate.Verify appears to handle IP addresses optionally surrounded by square brackets.
+		DNSName:       host,
+		Intermediates: intermediates,
+	}
+	roots, err := c.Verify(verificationOpts)
+
+	return len(roots) > 0, err
+}
```
Comment (clokep) [inline main.go:648]: This makes me a bit nervous, it seems it was removed in https://github.com/matrix-org/gomatrixserverlib/pull/364 and added in https://github.com/matrix-org/gomatrixserverlib/commit/5a698c3d2cc74f789296c8c198b38788aecbaf0a and never used anywhere (except here?)

tl;dr Is there a standard go way to validate the certificate instead of custom code like this?

File: main.go
Comment (clokep) [inline main.go:1]: Most of these changes are just from bumping the `gomatrixserverlib` used, I think?

File: main.go (hunk 1 above)
Comment (turt2live) [reply main.go:648]: fwiw, [here](https://github.com/turt2live/matrix-media-repo/blob/76dbb4b07397a461b02b71d3a5d9b847f2ab9172/matrix/federation.go#L254-L274) is how I do it in MMR (another Go project).

The block of code added in the PR here was mostly out of fear: it felt particularly magic, and I wasn't super interested in debugging *too much* of the federation tester 😅

File: main.go
Comment (turt2live) [reply main.go:1]: yes - should be its own commit

File: main.go [hunk 2]
```diff
@@ -360,31 +362,59 @@ func Report(
...
+		result.SRVCName, result.SRVRecords, err = net.LookupSRV("matrix-fed", "tcp", string(serverName))
 		result.SRVError = err
+		// Append the deprecated ones too
+		cname, records, err2 := net.LookupSRV("matrix", "tcp", string(serverName))
+		if result.SRVCName == "" {
+			result.SRVCName = cname
+		}
+		if records != nil {
+			result.SRVRecords = append(result.SRVRecords, records...)
+		}
```
Comment (clokep) [inline main.go:383]: Does this really make sense? I suspect it would be better to return it in a different field and update the UI to show reasonable information about the legacy vs. new ones.

Comment (turt2live) [reply main.go:383]: Returning a new field felt sufficiently complicated, for reasons I don't remember. I agree that a new field would be best.

File: main.go (hunk 1 above)
Comment (reivilibre) [reply main.go:648]: Given we're just doing this for informational purposes, I think this is OK. I mean, it also looks sane (but looks can be deceiving).
I can't see how you'd really knock any of this off — you need to be able to specify the DNS name and any intermediate certs. From recent experience working with X.509 I'd say it looks decent. But even if I was wrong, the worst we'll do is show the wrong result in the tester? I don't feel a grand need to worry too much.

File: main.go (hunk 2 above)
Comment (reivilibre) [reply main.go:383]: Will the UI tool be able to distinguish the two different types of SRV records based on this info?

I think a new field could possibly be best too, but either way I would like to ensure the information is there.

(Maybe you kind of want the result to be a warning of some sort if only the legacy option is set...)

**Review by reivilibre — APPROVED** [review]
overall looks OK to me. If you can address the field separation that would be good, otherwise it'd be nice to have a start at least I think, if the client UI can distinguish both cases anyway.

**Comment by turt2live** [comment]
This PR may need taking over from someone on the maintenance side, sorry. This PR is becoming fairly critical to deploy, but I'm running out of spare time to push it forward.

**Comment by 6543** [comment]
@turt2live `github.com/matrix-org/gomatrixserverlib` was bumped ... so this pull can have the simple addition without lib bumping now :)

File: main.go (hunk 1 above)
Comment (richvdh) [reply main.go]: This landed separately, in #155, in the end.
